domain/application/state/application.go

```diff
@@ -3287,3 +3323,77 @@ func decodeChannel(track string, risk sql.NullString, branch string) (*deploymen
 		Branch: branch,
 	}, nil
 }
+
+// getAllNonPeerRelationInfo retrieves metadata for all non-peer relations of a
+// given application.
+// It returns a slice of relationInfo containing all relevant information from
+// charm_relation and the count of application linked through this relation to
+// the current one.
+func (st *State) getAllNonPeerRelationInfo(ctx context.Context, tx *sqlair.TX, id coreapplication.ID) ([]relationInfo, error) {
+	type application dbUUID
+	app := application{UUID: id.String()}
+
+	stmt, err := st.Prepare(`
+SELECT
+    a.name AS &relationInfo.application_name,
+    vcr.charm_uuid AS &relationInfo.charm_uuid,
+    vcr.name AS &relationInfo.name,
+    vcr.role AS &relationInfo.role,
+    vcr.interface AS &relationInfo.interface,
+    vcr.optional AS &relationInfo.optional,
+    vcr.capacity AS &relationInfo.capacity,
+    vcr.scope AS &relationInfo.scope,
+    COUNT(DISTINCT re.relation_uuid) AS &relationInfo.count
+FROM   v_charm_relation AS vcr
+JOIN   application_endpoint AS ae ON vcr.uuid = ae.charm_relation_uuid    
+JOIN   relation_endpoint AS re ON ae.uuid = re.endpoint_uuid
+JOIN   application AS a ON ae.application_uuid = a.uuid
+WHERE  ae.application_uuid = $application.uuid
+AND    vcr.role != 'peer'
+GROUP BY a.name, vcr.charm_uuid, vcr.name, vcr.role, vcr.interface, vcr.optional, vcr.capacity, vcr.scope -- for count
+`, app, relationInfo{})
+	if err != nil {
+		return nil, errors.Errorf("preparing query: %w", err)
+	}
+
+	var result []relationInfo
+	if err := tx.Query(ctx, stmt, app).GetAll(&result); err != nil {
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return result, nil
+		}
+		return nil, errors.Errorf("failed to select non-peer relation info: %w", err)
+	}
+
+	return result, nil
+}
+
+func (st *State) precheckUpgradeRelation(meta *internalcharm.Meta, relations []relationInfo) error {
```

Comment:
todo: add a function header comment to list the criteria for refreshing a charm vis-a-vis relations.